src/condor_utils/proc_family_direct_cgroup_v1.cpp

```diff
@@ -0,0 +1,553 @@
+		if (!can_make_cgroup_dir) {
+			dprintf(D_ALWAYS, "Cannot mkdir %s, failing to use cgroups\n", absolute_cgroup.c_str());
+			return false;
+		}
+
+		// Now move pid to the leaf of the newly-created tree
+		stdfs::path procs_filename = absolute_cgroup / "cgroup.procs";
+		int fd = open(procs_filename.c_str(), O_WRONLY, 0666);
+		if (fd > 0) {
+			std::string buf;
+			formatstr(buf, "%u", pid);
+			int r = write(fd, buf.c_str(), strlen(buf.c_str()));
+			if (r < 0) {
+				dprintf(D_ALWAYS, "Error writing procid %d to %s: %s\n", pid, procs_filename.c_str(), strerror(errno));
+				close(fd);
+				return false;
+			} else {
+				dprintf(D_ALWAYS, "Moved process %d to cgroup %s\n", pid, absolute_cgroup.c_str());
+			}
+			close(fd);
+		}
+	}
+
+
+	// Set limits, if any
+	if (cgroup_memory_limit > 0) {
+		// write memory limits
+		stdfs::path memory_limits_path = cgroup_root_dir / "memory" / cgroup_name / "memory.limit_in_bytes";
+		int fd = open(memory_limits_path.c_str(), O_WRONLY, 0666);
+		if (fd > 0) {
+			std::string buf;
+			formatstr(buf, "%lu", cgroup_memory_limit);
+			int r = write(fd, buf.c_str(), strlen(buf.c_str()));
+			if (r < 0) {
+				dprintf(D_ALWAYS, "Error setting cgroup memory limit of %s in cgroup %s: %s\n", buf.c_str(), memory_limits_path.c_str(), strerror(errno));
+			}
+			close(fd);
+		} else {
+			dprintf(D_ALWAYS, "Error setting cgroup memory limit of %lu in cgroup %s: %s\n", cgroup_memory_limit, memory_limits_path.c_str(), strerror(errno));
+		}
+	} else {
+		dprintf(D_FULLDEBUG, "ProcFamilyDirectCgroupV1 not setting any cgroup memory limits\n");
+	}
+
+	// Set cpu limits, if any
+	if (cgroup_cpu_shares > 0) {
+		// write CPU limits
+		stdfs::path cpu_shares_path = cgroup_root_dir / "cpu,cpuacct" / cgroup_name / "cpu.shares";
+		int fd = open(cpu_shares_path.c_str(), O_WRONLY, 0666);
+		if (fd > 0) {
+			std::string buf;
+			formatstr(buf, "%d", cgroup_cpu_shares);
+			int r = write(fd, buf.c_str(), buf.length());
+			if (r < 0) {
+				dprintf(D_ALWAYS, "Error setting cgroup cpu weight of %d in cgroup %s: %s\n", cgroup_cpu_shares, cpu_shares_path.c_str(), strerror(errno));
+			}
+			close(fd);
+		} else {
+			dprintf(D_ALWAYS, "Error setting cgroup cpu weight of %d in cgroup %s: %s\n", cgroup_cpu_shares, cpu_shares_path.c_str(), strerror(errno));
+		}
+	}
+
+	// Chown the leaf cgroup dirs to the non-root owner.  The control files wlil
+	// still be owned by root, so the job can't change them, but they can
+	// create sub-dirs, so that a glidein can manage the resources given in this
+	// cgroup
+	pid_t uid = get_user_uid();
+	pid_t gid = get_user_gid();;
+
+	// in a personal condor, uid/gid aren't returned properly.
+	if ((uid > 0) && (gid > 0)) {
+		for (auto &controller: controllers) {
+			int r = chown((cgroup_root_dir / controller / cgroup_name).c_str(), uid, gid);
+			if (r < 0) {
+				dprintf(D_FULLDEBUG, "Error chowning cgroup directory: %s to (%d.%d)\n", strerror(errno), uid, gid);
```

Comment:
I guess this error will be detected when we try doing something later?

Reply:
Yes -- at the worst it falls back to today's behaviour. And in the usual case where there isn't a glidein running in us, there's no difference.